File: core/src/main/java/com/safframework/rxcache/rxjava3/transformstrategy/impl/CacheAndRemoteStrategy.java
package com.safframework.rxcache.rxjava3.transformstrategy.impl;

import com.safframework.rxcache.RxCache;
import com.safframework.rxcache.domain.Record;
import com.safframework.rxcache.domain.Source;
import com.safframework.rxcache.rxjava3.transformstrategy.FlowableStrategy;
import com.safframework.rxcache.rxjava3.transformstrategy.MaybeStrategy;
import com.safframework.rxcache.rxjava3.transformstrategy.ObservableStrategy;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.BackpressureStrategy;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.functions.Function;
import io.reactivex.rxjava3.functions.Predicate;
import org.reactivestreams.Publisher;

import java.lang.reflect.Type;
import java.util.Arrays;

/**
 * 先获取缓存，再获取网络请求
 * Created by tony on 2019-01-25.
 */
public class CacheAndRemoteStrategy implements ObservableStrategy,
        FlowableStrategy,
        MaybeStrategy {


    @Override
    public <T> Publisher<Record<T>> execute(RxCache rxCache, String key, Flowable<T> source, Type type) {

        Flowable<Record<T>> cache = rxCache.<T>load2Flowable(key, type);

        Flowable<Record<T>> remote = source
                .map(new Function<T, Record<T>>() {
                    @Override
                    public Record<T> apply(@NonNull T t) throws Exception {

                        rxCache.save(key, t);

                        return new Record<>(Source.CLOUD, key, t);
                    }
                });

        return Flowable.concatDelayError(Arrays.asList(cache, remote))
                .filter(new Predicate<Record<T>>() {
                    @Override
                    public boolean test(@NonNull Record<T> record) throws Exception {
                        return record.getData() != null;
                    }
                });
    }

    @Override
    public <T> Publisher<Record<T>> execute(RxCache rxCache, String key, Flowable<T> source, Type type, BackpressureStrategy backpressureStrategy) {

        Flowable<Record<T>> cache = rxCache.<T>load2Flowable(key, type, backpressureStrategy);

        Flowable<Record<T>> remote = source
                .map(new Function<T, Record<T>>() {
                    @Override
                    public Record<T> apply(@NonNull T t) throws Exception {

                        rxCache.save(key, t);

                        return new Record<>(Source.CLOUD, key, t);
                    }
                });

        return Flowable.concatDelayError(Arrays.asList(cache, remote))
                .filter(new Predicate<Record<T>>() {
                    @Override
                    public boolean test(@NonNull Record<T> record) throws Exception {
                        return record.getData() != null;
                    }
                });
    }

    @Override
    public <T> Maybe<Record<T>> execute(RxCache rxCache, String key, Maybe<T> source, Type type) {

        Maybe<Record<T>> cache = rxCache.<T>load2Maybe(key, type);

        Maybe<Record<T>> remote = source
                .map(new Function<T, Record<T>>() {
                    @Override
                    public Record<T> apply(@NonNull T t) throws Exception {

                        rxCache.save(key, t);

                        return new Record<>(Source.CLOUD, key, t);
                    }
                });

        return Maybe.concatDelayError(Arrays.asList(cache,remote))
                .filter(new Predicate<Record<T>>() {
                    @Override
                    public boolean test(@NonNull Record<T> record) throws Exception {
                        return record.getData() != null;
                    }
                })
                .firstElement();
    }

    @Override
    public <T> Observable<Record<T>> execute(RxCache rxCache, String key, Observable<T> source, Type type) {

        Observable<Record<T>> cache = rxCache.<T>load2Observable(key, type);

        Observable<Record<T>> remote = source
                .map(new Function<T, Record<T>>() {
                    @Override
                    public Record<T> apply(@NonNull T t) throws Exception {

                        rxCache.save(key, t);

                        return new Record<>(Source.CLOUD, key, t);
                    }
                });

        return Observable.concatDelayError(Arrays.asList(cache, remote))
                .filter(new Predicate<Record<T>>() {
                    @Override
                    public boolean test(@NonNull Record<T> record) throws Exception {
                        return record.getData() != null;
                    }
                });
    }
}

File: rxjava/rxjava2/src/main/kotlin/com/safframework/rxcache/rxjava/rxjava2/impl/CacheAndRemoteStrategy.kt
package com.safframework.rxcache.rxjava.rxjava2.impl

import com.safframework.rxcache.RxCache
import com.safframework.rxcache.domain.Record
import com.safframework.rxcache.domain.Source
import com.safframework.rxcache.rxjava.rxjava2.load2Flowable
import com.safframework.rxcache.rxjava.rxjava2.load2Maybe
import com.safframework.rxcache.rxjava.rxjava2.load2Observable
import com.safframework.rxcache.rxjava.rxjava2.transformstrategy.FlowableStrategy
import com.safframework.rxcache.rxjava.rxjava2.transformstrategy.MaybeStrategy
import com.safframework.rxcache.rxjava.rxjava2.transformstrategy.ObservableStrategy
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.Maybe
import io.reactivex.Observable
import org.reactivestreams.Publisher
import java.lang.reflect.Type
import java.util.*

/**
 *
 * @FileName:
 *          com.safframework.rxcache.rxjava.rxjava3.transformstrategy.impl.CacheAndRemoteStrategy
 * @author: Tony Shen
 * @date: 2021-02-19 13:06
 * @version: V1.0 先获取缓存，再获取网络请求
 */
class CacheAndRemoteStrategy : ObservableStrategy, FlowableStrategy, MaybeStrategy {

    override fun <T> execute(rxCache: RxCache, key: String, source: Flowable<T>, type: Type): Publisher<Record<T>> {
        val cache: Flowable<Record<T>> = rxCache.load2Flowable(key, type)
        val remote: Flowable<Record<T>> = source
            .map{ t ->
                rxCache.save(key, t)
                Record(Source.CLOUD, key, t)
            }
        return Flowable.concatDelayError(Arrays.asList(cache, remote))
            .filter{ record -> record.data != null }
    }

    override fun <T> execute(
        rxCache: RxCache,
        key: String,
        source: Flowable<T>,
        type: Type,
        backpressureStrategy: BackpressureStrategy
    ): Publisher<Record<T>> {
        val cache: Flowable<Record<T>> = rxCache.load2Flowable(key, type, backpressureStrategy)
        val remote: Flowable<Record<T>> = source
            .map{ t ->
                rxCache.save(key, t)
                Record(Source.CLOUD, key, t)
            }
        return Flowable.concatDelayError(Arrays.asList(cache, remote))
            .filter{ record -> record.data != null }
    }

    override fun <T> execute(rxCache: RxCache, key: String, source: Maybe<T>, type: Type): Maybe<Record<T>> {
        val cache: Maybe<Record<T>> = rxCache.load2Maybe(key, type)
        val remote: Maybe<Record<T>> = source
            .map{ t ->
                rxCache.save(key, t)
                Record(Source.CLOUD, key, t)
            }
        return Maybe.concatDelayError(Arrays.asList(cache, remote))
            .filter{ record -> record.data != null }
            .firstElement()
    }

    override fun <T> execute(rxCache: RxCache, key: String, source: Observable<T>, type: Type): Observable<Record<T>> {
        val cache: Observable<Record<T>> = rxCache.load2Observable(key, type)
        val remote: Observable<Record<T>> = source
            .map{ t ->
                rxCache.save(key, t)
                Record(Source.CLOUD, key, t)
            }
        return Observable.concatDelayError(Arrays.asList(cache, remote))
            .filter{ record -> record.data != null }
    }
}